pkg/controller/workload/kubernetes/resource/resource.go

```diff
@@ -0,0 +1,577 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	computev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/compute/v1alpha1"
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/apis/workload/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/controller/core"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+	"github.com/crossplaneio/crossplane/pkg/util"
+)
+
+const (
+	controllerName   = "kubernetesapplicationresource." + v1alpha1.Group
+	finalizerName    = "finalizer." + controllerName
+	reconcileTimeout = 1 * time.Minute
+
+	reasonFetchingClient   = "failed to fetch Kubernetes client"
+	reasonSyncingResource  = "failed to sync templated resource in Kubernetes cluster"
+	reasonDeletingResource = "failed to delete templated resource from Kubernetes cluster"
+	reasonGettingSecret    = "failed to get connection secret for resource dependency"
+	reasonSyncingSecret    = "failed to update connection secret for resource dependency"
+	reasonDeletingSecret   = "failed to delete connection secret for resource dependency"
+
+	messageMissingTemplate = v1alpha1.KubernetesApplicationResourceKind + " must include a template"
+)
+
+// Ownership annotations
+const (
+	RemoteControllerNamespace = v1alpha1.KubernetesApplicationResourceKind + "." + v1alpha1.Group + "/namespace"
+	RemoteControllerName      = v1alpha1.KubernetesApplicationResourceKind + "." + v1alpha1.Group + "/name"
+	RemoteControllerUID       = v1alpha1.KubernetesApplicationResourceKind + "." + v1alpha1.Group + "/uid"
+)
+
+var (
+	log = logging.Logger.WithName("controller." + controllerName)
+)
+
+// CreatePredicate accepts KubernetesApplicationResources that have been
+// scheduled to a KubernetesCluster.
+func CreatePredicate(event event.CreateEvent) bool {
+	wl, ok := event.Object.(*v1alpha1.KubernetesApplicationResource)
+	if !ok {
+		return false
+	}
+	return wl.Status.Cluster != nil
```

Comment:
`wl.Status.Cluster` is `nil` if the `KubernetesApplicationResource` is 'pending scheduling'.

This is (kind of) the same pattern as the `Status.Cluster` field for `KubernetesApplication` and the existing `Workload` types. The difference is that there's no scheduler that operates directly on a `KubernetesApplicationResource` - the only way a `KubernetesApplicationResource's` `Status.Cluster` can be set set is by the `KubernetesApplication` controller.

That said, there's nothing stopping a Crossplane user from submitting a `KubernetesApplicationResource` by itself, rather than as part of a `KubernetesApplication`, so these predicates are intended to prevent us trying to reconcile these potential, 'unschedulable' `KubernetesApplicationResources`.

Reply:
This brings up a related thought - I feel that it should be possible for a Crossplane user to create a functional `KubernetesApplicationResource` directly, i.e. without a using `KubernetesApplication`.

This is not because I think there are obvious practical use cases in which a Crossplane user would want to create and manage `KubernetesApplicationResources` manually (though maybe there could be?), but rather because it seems to go against Kubernetes convention to add a resource kind that is only useful when created and managed by another resource kind. A good example here is `ReplicaSet`. In 99% of cases they'll be created and managed by a `Deployment` - there's relatively few use cases for creating and managing them manually - but it's still _possible_ to do so.

I could see two ways to address this:
1. Add a scheduler that schedules `KubernetesApplicationResources` with `nil` clusters. I expect this scheduler would see very rare use, given that the `KubernetesApplication` controller creates "pre-scheduled" `KubernetesApplicationResources`.
1. Move the `Cluster` field from `Status` into `Spec` so a user could explicitly specify where they wanted their `KubernetesApplicationResource` to run.

I prefer option one - said scheduler would likely share a lot with the existing `KubernetesApplication` scheduler and thus hopefully not be too much of a maintenance burden.

Reply:
Thank you for clarifying this. Yes, this was my initial concert. I noticed that scheduler did not support `KubernetesApplicatoinResource` objects, hence it was not clear how/when the Cluster assignment would happen.

Reply:
I'll implement this scheduler in a separate PR later this week.